src/ipc/server.rs

```diff
@@ -208,6 +208,21 @@ async fn process(ctx: &ClientCtx, request: Request) -> Reply {
             let workspaces = result.map_err(|_| String::from("error getting workspace info"))?;
             Response::Workspaces(workspaces)
         }
+        Request::FocusedOutput => {
+            let (tx, rx) = async_channel::bounded(1);
+            ctx.event_loop.insert_idle(move |state| {
+                let active_output = state
+                    .niri
+                    .layout
+                    .active_output()
+                    .map(|output| output.name());
+                let _ = tx.send_blocking(active_output);
+            });
+            let result = rx.recv().await;
+            let name = result.map_err(|_| String::from("error getting active output info"))?;
+            let output = name.and_then(|name| ctx.ipc_outputs.lock().unwrap().get(&name).cloned());
```

Comment:
Do this inside the `insert_idle` closure to guarantee that the output exists. Otherwise it can get disconnected between when the idle runs, and when this line runs. Instead of `ctx.ipc_outputs` you can extract it from `state.niri.global_space.outputs()`.

Reply:
i have been looking at this for a little bit and the `global_space.outputs()` returns smithay `Output`s instead of `niri_ipc::Output`s. what am i missing?

Reply:
Oh, right. Sorry, got confused. Then get it from the arc-mutex, but within the closure. The arc-mutex should be available somewhere in the backend or in the niri state.

Reply:
that is what i ended up doing, thanks